### PR TITLE
Tests assets bundles

### PR DESCRIPTION
**Pull request by Polymorphe57**

### CODE REVIEW

File: odoo/addons/test_assetsbundle/controllers/main.py [hunk 1]
```diff
@@ -8,6 +8,4 @@ class TestAssetsBundleController(Controller):
     @route('/test_assetsbundle/js', type='http', auth='user')
     def bundle(self):
         env = request.env(user=SUPERUSER_ID)
-        bundle = env.ref('test_assetsbundle.bundle1')
-        views = env['ir.ui.view'].search([('inherit_id', '=', bundle.id)])
-        return views.with_context(check_view_ids=views.ids)._render_template('test_assetsbundle.template1')
+        return env['ir.ui.view']._render_template('test_assetsbundle.template1')
```
Comment (Polymorphe57) [inline odoo/addons/test_assetsbundle/controllers/main.py:11]: I am not sure this replaces the old test correctly

File: odoo/addons/base/models/ir_qweb.py [hunk 2]
```diff
@@ -320,7 +320,9 @@ def get_file_info(file):
                 'media': None,
             }
 
-        addons = module_boot()
+        addons = []
```
Comment (Polymorphe57) [inline odoo/addons/base/models/ir_qweb.py]: we have a problem before no need to have addons. We just used to  access read a particular ir.ui.view using its xml id. For now we can only make test pass if we create an ir.asset record. Bundles created via the manifest don't work when we are not in a request

Comment (Polymorphe57) [reply odoo/addons/base/models/ir_qweb.py]: in previous implementation of _get_asset_content
module_boot was only useful to set get_modules_order in template rendering context and it was only set when request was set

File: odoo/addons/base/models/ir_qweb.py [hunk 3]
```diff
@@ -314,7 +314,9 @@ def _get_asset_content(self, xmlid, options, css=True, js=True):
         IrAsset = self.env['ir.asset']
 
         addons = []
-        if request:
+        if not request:
+            addons = self.env['ir.module.module']._installed_sorted()
```
Comment (Polymorphe57) [inline odoo/addons/base/models/ir_qweb.py]: why not always this (request or not)?

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py [hunk 4]
```diff
@@ -632,3 +636,366 @@ def test_01_debug_mode_assets(self):
...
+        ''' % {'asset_key': asset_key}
+
+        view = self.env['ir.ui.view'].create({
+            'name': 'test asset',
+            'arch': arch,
+            'type': 'qweb',
+        })
+        return view
+
+    def assertStringEqual(self, reference, tested):
```
Comment (Polymorphe57) [inline odoo/addons/test_assetsbundle/tests/test_assetsbundle.py:831]: tested <-> reference

File: odoo/addons/base/models/ir_qweb.py [hunk 5]
```diff
@@ -312,7 +312,12 @@ def _get_asset_content(self, xmlid, options, css=True, js=True):
 
         options['website_id'] = self.env.context.get('website_id')
         IrAsset = self.env['ir.asset']
-        addons = module_boot()
+
+        addons = []
+        if not request:
```
Comment (Polymorphe57) [inline odoo/addons/base/models/ir_qweb.py]: could we use  always _installed_sorted?

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py [hunk 6]
```diff
@@ -632,3 +660,340 @@ def test_01_debug_mode_assets(self):
...
+            var b=2;;
+
+            /* /test_assetsbundle/static/src/js/test_jsfile3.js defined in bundle 'test_assetsbundle.manifest1' */
+            var c=3;;
+
+            /* /test_assetsbundle/static/src/js/test_jsfile4.js defined in bundle 'test_assetsbundle.manifest1' */
+            var d=4;
+            '''
+        )
+        # LPE FIXME: where is my http glob ?
```
Comment (kebeclibre) [inline odoo/addons/test_assetsbundle/tests/test_assetsbundle.py]: @Arcasias 
Hey, There are a few blindspots in those tests, including this line.
Do you mind checking it to refine the specs ?

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py [hunk 7]
```diff
@@ -632,3 +660,340 @@ def test_01_debug_mode_assets(self):
...
+        view = self.make_asset_view('test_assetsbundle.manifest2')
+        self.env['ir.asset'].create({
+            'name': 'test_jsfile4',
+            'bundle': 'test_assetsbundle.manifest2',
+            'directive': 'remove',
+            'glob': 'test_assetsbundle/static/src/*/**',
+        })
+        view._render()
+        attach = self.env['ir.attachment'].search([('name', 'ilike', 'test_assetsbundle.manifest2.js')], order='create_date DESC', limit=1)
+        # LPE Fixme: is this really what we want ?
```
Comment (kebeclibre) [inline odoo/addons/test_assetsbundle/tests/test_assetsbundle.py]: @Arcasias  here too

**Review by kebeclibre — COMMENTED** [review]
@ged-odoo 
@aab-odoo 
Here is what I found to be worth testing.
Can you think of more twisted use cases ?

**Comment by aab-odoo** [comment]
@kebeclibre I didn't see any test about commands except for `include`? I may be outdated spec-wize, but as far as I can remember, we have `append`, `prepend`, `replace` (and maybe more?). They should be tested :-)

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py (hunk 6 above)
Comment (Arcasias) [reply odoo/addons/test_assetsbundle/tests/test_assetsbundle.py]: This is the same behaviour as the previous system: URLs are not aggregated in assets bundles, they are instead appended as is to the arch where `t-call-assets` was called (see `TestJavascriptAssetsBundle.test_20_exteral_lib_assets`). Now I should perhaps make that point clear in the doc if it wasn't before.

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py (hunk 7 above)
Comment (Arcasias) [reply odoo/addons/test_assetsbundle/tests/test_assetsbundle.py]: It doesn't seem to be working here... I would've expected the whole `src` directory to be wiped from the resulting assets. Is it perhaps because you used `*/**` instead of `**/*`? Not being an expert in globs I don't even know if there is a difference

**Comment by kebeclibre** [comment]
drop node attribute on external links

media= & direction= should be handled at the t-call-asset step
should test in master

File: addons/survey/views/assets.xml [hunk 8]
```diff
@@ -17,7 +17,6 @@
         <t t-call="web._assets_helpers">
             <t t-call="web._assets_frontend_helpers"/>
         </t>
-
```
Comment (Polymorphe57) [inline addons/survey/views/assets.xml]: jum has added a blank line in his branch

**Comment by kebeclibre** [comment]
@Arcasias 
@aab-odoo 
@ged-odoo 
Tests for asset feature are ready. Don't hesitate to review them and tell if you could think of twisted use cases.

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py [hunk 9]
```diff
@@ -572,11 +576,38 @@ def test_02_js_interpretation_inline(self):
...
+    # LPE Fixme
+    # Review point @al: is this really what we want people to do ?
+    def test_03_js_interpretation_recommended_new_method(self):
+        """ Checks the feature of test_02 is still produceable, but in another way
+        '/web/content/<int:id>/<string: filename.js>',
+        """
+        code = b'const d = 4;'
+        attach = self.env['ir.attachment'].create({
+            'name': 'CustomJscode.js',
+            'mimetype': 'text/javascript',
```
Comment (brboi) [inline odoo/addons/test_assetsbundle/tests/test_assetsbundle.py:722]: application/javascript ?

File: odoo/addons/base/models/ir_qweb.py [hunk 10]
```diff
@@ -295,45 +298,71 @@ def get_asset_bundle(self, xmlid, files, env=None):
...
-        files, remains = self._get_asset_content(xmlid, options, css, js)
+    def _get_asset_nodes(self, xmlid, options, css=True, js=True, debug=False, async_load=False, defer_load=False, lazy_load=False, media=None, values=None):
+        nodeAttrs = None
+        if css and media:
+            nodeAttrs = {
+                'media': media,
+            }
+        files, remains = self._get_asset_content(xmlid, options, nodeAttrs)
         asset = self.get_asset_bundle(xmlid, files, env=self.env)
+        remains = [node for node in remains if (css and node[0] == 'link') or (js and node[0] == 'script')]
```
Comment (Arcasias) [inline odoo/addons/base/models/ir_qweb.py:320]: Is this relevant? All paths and URLs should already be filtered in `get_addon_files`

File: odoo/addons/base/models/ir_qweb.py [hunk 11]
```diff
@@ -295,45 +298,71 @@ def get_asset_bundle(self, xmlid, files, env=None):
...
+
+        addon_files = self.env['ir.asset'].get_addon_files(addons=addons, bundle=xmlid, css=True, js=True)
 
         files = []
         remains = []
         for _, file in addon_files:
+            ext = file.split('.')[-1]
+            is_js = ext in SCRIPT_EXTENSIONS
+            is_css = ext in STYLE_EXTENSIONS
+            if not is_js and not is_css:
```
Comment (Arcasias) [inline odoo/addons/base/models/ir_qweb.py:344]: Same here: if `js=True` the `get_addon_files` method should only return JS paths/urls, and the same applies to `css=True`. If this is necessary I guess there is a problem to fix in `ir_asset.py` (or I'm just missing something).

File: odoo/addons/base/models/ir_qweb.py (hunk 10 above)
Comment (kebeclibre) [reply odoo/addons/base/models/ir_qweb.py:320]: @Arcasias 
I believe it is relevant, but not linked to your algo by itself.

TLDR: for performance reasons, we ormcahe a max of stuff, we need to go your algo only once, with both js and css to true; the heuristic on the file extension is then necessary

What was done before, and that we tried to reproduce (this is a performance decision that needs to be disputed :D ) is that we go through the asset only once (we go through your algo only once) with css and js to True.

The main parent function call to your algo is _get_asset_nodes ; which is orrmcached on the booleans css and js. 

Since _get_asset_content is not ormcached on js and css booleans, it needs to have done the filtering work for both css and js at the same time, hence the heuristic on the file extension.

_get_asset_nodes takes the result of _get_asset_content (which is now a constant containing both js and css) and just has to apply the "remain" filter on it (twice : one for js and one for css)

Comment (Arcasias) [reply odoo/addons/base/models/ir_qweb.py:320]: Right... did not see the `js` and `css` arguments both to true :D
Alright then, fair point.
